AX: Trigger timeout in 'waitFor' function if promise could not be resolved

dpino opened the pull request:
#### 1e32f01796045db9a88ba75f83ab6239f468a16d
<pre>
AX: Trigger timeout in &apos;waitFor&apos; function if promise could not be resolved
<a href="https://bugs.webkit.org/show_bug.cgi?id=298590">https://bugs.webkit.org/show_bug.cgi?id=298590</a>

Reviewed by Tyler Wilcock.

When evaluating an expression in &apos;expectAsync&apos; or &apos;expectAsyncExpression&apos;, if
the expression evaluated is false or triggers an exception, the operator &apos;await&apos;
will hang waiting for the promise to be resolved or rejected. As a consequence
of that, many subtests that may fail are actually timing out.

To prevent &apos;waitFor&apos; to hang forever, a timeout timer was added. The timer will
resolve the promise if it could not be resolved within a time threshold
(3 seconds).

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/resources/accessibility-helper.js:

Canonical link: <a href="https://commits.webkit.org/299867@main">https://commits.webkit.org/299867@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1883128d9893548e72433ccf1050e248a25c0c1c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120468 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/40162 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30814 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126844 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72543 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40859 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48739 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91494 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60776 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123420 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32629 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107985 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72045 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31665 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26092 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/70460 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102113 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26273 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129729 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47389 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35964 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100114 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47755 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104167 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99956 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45398 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23430 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/44037 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19128 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/47251 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46719 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50066 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48406 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f4c953da08b60c062787c18288dbdb88c497f4df. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50487

dpino committed:
AX: Trigger timeout in 'waitFor' function if promise could not be resolved
https://bugs.webkit.org/show_bug.cgi?id=298590

Reviewed by Tyler Wilcock.

When evaluating an expression in 'expectAsync' or 'expectAsyncExpression', if
the expression evaluated is false or triggers an exception, the operator 'await'
will hang waiting for the promise to be resolved or rejected. As a consequence
of that, many subtests that may fail are actually timing out.

To prevent 'waitFor' to hang forever, a timeout timer was added. The timer will
resolve the promise if it could not be resolved within a time threshold
(3 seconds).

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/resources/accessibility-helper.js:

Canonical link: https://commits.webkit.org/299867@main
